File: Linked_Lists/python/duplicates.py
from LinkedList import  LinkedList, Node
from random import randint


def remove_duplicates1(linked_list: LinkedList):
    nums = []
    offset = 0

    for i, node in enumerate(linked_list):
        if node.data in nums:
            linked_list.remove_at_index(i - offset)
            offset += 1
        else:
            nums.append(node.data)

    return linked_list


def remove_duplicates2(linked_list: LinkedList):
    nums = []

    for i, node in enumerate(linked_list):
        if node.data in nums:
            prev.next = node.next
        else:
            nums.append(node.data)
            prev = node

    return linked_list


def remove_duplicates3(linked_list: LinkedList):
    current = linked_list.head

    while current.next:
        runner = current
        while runner.next:
            if current.data == runner.next.data:
                runner.next = runner.next.next
            else:
                runner = runner.next

        current = current.next

    return linked_list


def main():
    linked_list = LinkedList()
    for i in range(10):
        linked_list.append(randint(0, 5))

    print(linked_list)

    print(remove_duplicates1(linked_list))
    print(remove_duplicates3(linked_list))


if __name__ == '__main__':
    main()

File: Linked_Lists/go/duplicates.go
package main

import (
  // "fmt"
  "types"
  "math/rand"
)


func valueInArray(array []int, value int) bool {
  if len(array) == 0 {
    return false
  }

  for _, num := range array {
    if num == value {
      return true
    }
  }
  return false
}


func removeDuplicates(list types.LinkedList) {
  nums := make([]int, 0)
  node := list.Head
  prev := node

  for node != nil {
    if valueInArray(nums, node.Data) {
      prev.Next = node.Next
    } else {
      nums = append(nums, node.Data)
      prev = node
    }
    node = node.Next
  }

}


func main() {
  var list types.LinkedList
  var length int = 5

  for i := 0; i < length; i++ {
    list.Append(rand.Intn(23))
  }

  list.Append(0)
  list.Append(1)
  list.Append(1)
  list.Append(22)
  list.Append(4)
  list.Append(5)
  list.Append(4)

  types.PrintLinkedList(list)
  removeDuplicates(list)
  types.PrintLinkedList(list)
}
